ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/services/transaction/TransactionConversion.scala

```diff
@@ -3,64 +3,111 @@
+  def acceptedToDomainFlat(
+      trans: AcceptedTransaction,
+      filter: domain.TransactionFilter
+  ): Option[domain.Transaction] = {
+    val tx = trans.transaction.mapNodeId(domain.EventId(_))
+    val events = engine.Event
+      .collectEvents(tx, trans.explicitDisclosure.map { case (k, v) => domain.EventId(k) -> v })
+    val allEvents = events.roots.toSeq
+      .sortBy(evId => getEventIndex(evId.unwrap))
+      .foldLeft(List.empty[CreateOrArchiveEvent])((l, evId) =>
+        l ::: flattenEvents(events.events, evId, true))
+
+    val eventFilter = TemplateAwareFilter(filter)
+    val filteredEvents = TransientContractRemover
+      .removeTransients(allEvents)
+      .flatMap(eventFilter.filterCreateOrArchiveWitnesses(_).toList)
+
+    val submitterIsSubscriber =
+      trans.submitter
+        .map(LfRef.Party.assertFromString)
+        .fold(false)(eventFilter.isSubmitterSubscriber)
+    if (filteredEvents.nonEmpty || submitterIsSubscriber) {
+      Some(
+        domain.Transaction(
+          domain.TransactionId(trans.transactionId),
+          Tag.subst(trans.commandId).filter(_ => submitterIsSubscriber),
+          Tag.subst(trans.workflowId),
+          trans.recordTime,
+          filteredEvents,
+          domain.LedgerOffset.Absolute(trans.offset),
+          None
+        ))
+    } else None
   }
-  private def convert(ps: Set[LfRef.Party]): Seq[String] = ps.toSeq
-
-  def eventsToTransaction(
-      allEvents: P.Events[EventId, AbsoluteContractId],
-      verbose: Boolean): TransactionTreeNodes = {
-    val events = allEvents.events.map {
-      case (nodeId, value) =>
-        (nodeId, value match {
-          case e: P.ExerciseEvent[EventId, AbsoluteContractId] =>
-            TreeEvent(TreeEvent.Kind.Exercised(lfExerciseToApi(nodeId, e, verbose)))
-          case c: P.CreateEvent[AbsoluteContractId] =>
-            TreeEvent(TreeEvent.Kind.Created(lfCreateToApi(nodeId, c, true, verbose)))
-        })
-    }
 
-    removeInvisibleRoots(events, allEvents.roots.toList.sortBy(getEventIndex))
+  def acceptedToDomainTree(
+      trans: AcceptedTransaction,
+      filter: domain.TransactionFilter): Option[domain.TransactionTree] = {
+
+    val tx = trans.transaction.mapNodeId(domain.EventId(_))
+    filter.filter(tx).map { disclosureByNodeId =>
+      val allEvents = engine.Event
+        .collectEvents(tx, disclosureByNodeId)
+      val events = allEvents.events.map {
+        case (nodeId, value) =>
+          (nodeId, value match {
+            case e: P.ExerciseEvent[domain.EventId, AbsoluteContractId] =>
+              lfExerciseToDomain(nodeId, e)
+            case c: P.CreateEvent[AbsoluteContractId] =>
+              lfCreateToDomain(nodeId, c, true)
+          })
+      }
+
+      val (byId, roots) =
+        removeInvisibleRoots(
+          events,
+          allEvents.roots.toList
+            .sortBy(evid => getEventIndex(evid.unwrap)))
+      val subscriberIsSubmitter = trans.submitter.forall(sub =>
```

Comment:
@gaboraranyossy-da: the previous code that corresponds to the `forall` is in [VisibleTransaction.scala](https://github.com/digital-asset/daml/blob/master/ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/services/transaction/VisibleTransaction.scala#L42-L45), where keep the commandId in case the submitter is unknown.
Do we expect in this case that the command id would also already be `None`, or should we explicitly only keep the commandId if the subscriber is also the submitter?

Reply:
It looks to me that previously the flat transaction response and the transaction tree response were inconsistent in this regard.
[FlatTransaction behavior](https://github.com/digital-asset/daml/pull/1404/files#diff-f81f736de2b1a109b981cf125fb7b4deL113): `.fold(false)(...)` == `.exists(...)`
[TransactionTree behavior](https://github.com/digital-asset/daml/blob/master/ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/services/transaction/VisibleTransaction.scala#L42-L45): `case _ => transaction.meta` == not changing the data if `transaction.meta.submitter == None` == ".forall(...)"

Reply:
Clarified offline: both submitter and commandId of `AcceptedTransaction` are set to `None` if the ledger node does not host the submitting party (i.e. retrieved the transaction via the up-and-coming ReadService). For sake of consistency, we'll use `exists` like we do for flat transactions.